### PR TITLE
Cosim coupl monitor

### PR DESCRIPTION
**Pull request by dionperd**
Added some property methods in CoSimulator class to compute existing_connections, min_idelay and min_delay.
Added also relative_output_time_steps to be able to get delayed outputs from monitors from further in the past, e.g., from 2*synchronization_time in the past.

Modified the CosimCouplingMonitor to be able to provide future values up to min_idelay steps instead of synchronization_n_step.

Modified the CosimHistory to have a length of TVB min_idelay instead of synchronization_n_step.

Allowing negative relative time steps for the loop_cosim_monitor_output method of CoSimunlator, in order to be able to send the TVB state or the initial condition to the other co-simulator from further in the past, , e.g., from 2*synchronization_time in the past.

This was necessary to allow for sending to the transformers future values of the period 
[current_step + min_idelay/2, current_step+min_idelay] 
to be processed  in parallel with current cosimulation
 [current_step, current_step + min_idelay/2],
 assuming that synchronization_n_step = min_idelay/2.

This will work in settings where Transformers can run in parallel with the CoSimulators.

All cosimulation tests PASS.

### CODE REVIEW

File: tvb_contrib/tvb/contrib/cosimulation/cosimulator.py [hunk 1]
```diff
@@ -312,9 +331,8 @@ def loop_cosim_monitor_output(self, n_steps=None, relative_start_step=0):
                 # check if it's valid input
                 ValueError("Incorrect n_steps = %i; it should be <= %i".format(
                            n_steps, self.good_cosim_update_values_shape[0]))
-            if relative_start_step < 0 or \
```
Comment (lionelkusch) [inline tvb_contrib/tvb/contrib/cosimulation/cosimulator.py:315]: Why did you remove this condition?

File: tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py [hunk 2]
```diff
@@ -97,15 +97,19 @@ class CosimMonitorFromCoupling(CosimMonitor):
                dynamic equations of the Model. Its primary purpose is to 'rescale' the
                incoming activity to a level appropriate to Model.""")
 
-    synchronization_n_step = None
+    min_idelay = Int(
```
Comment (lionelkusch) [inline tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py:100]: You should not define it as a parameter because the value defined by the user won't be used. 
min_idelay should be defined as a private variable, i.e. _min_idelay without defining the type.

File: tvb_contrib/tvb/contrib/cosimulation/cosimulator.py [hunk 3]
```diff
@@ -78,6 +78,27 @@ class CoSimulator(Simulator):
     _cosim_monitors_noncoupling_indices = []
     _cosim_monitors_coupling_indices = []
 
+    @property
```
Comment (lionelkusch) [inline tvb_contrib/tvb/contrib/cosimulation/cosimulator.py:81]: Using "@property" cannot be efficient in some situations because it is computed each time it is called.
For the moment, there are few calls to it, so it should be fine.

File: tvb_contrib/tvb/contrib/cosimulation/cosimulator.py [hunk 4]
```diff
@@ -90,14 +111,12 @@ def _configure_synchronization_time(self):
...
         # Check if the synchronization time is smaller than the minimum delay of the connectivity:
-        existing_connections = self.connectivity.weights != 0
-        if numpy.any(existing_connections):
-            min_idelay = self.connectivity.idelays[existing_connections].min()
-            if self.synchronization_n_step > min_idelay:
-                raise ValueError('The synchronization time %g is longer than '
-                                 'the minimum delay time %g '
-                                 'of all existing connections (i.e., of nonzero weight)!'
-                                 % (self.synchronization_time, min_idelay * self.integrator.dt))
+        min_idelay = self.min_idelay
```
Comment (lionelkusch) [inline tvb_contrib/tvb/contrib/cosimulation/cosimulator.py:114]: Merge this line with the following line.

**Review by lionelkusch — CHANGES_REQUESTED** [review]
Thank you for relaxing the condition of the co-simulation monitor. 
I have some remarks to fix before merging.

File: tvb_contrib/tvb/contrib/cosimulation/cosimulator.py (hunk 1 above)
Comment (dionperd) [reply tvb_contrib/tvb/contrib/cosimulation/cosimulator.py:315]: > Why did you remove this condition?

Allowing negative relative time steps for the loop_cosim_monitor_output method of CoSimunlator, in order to be able to send the TVB state or the initial condition to the other co-simulator from further in the past, , e.g., from 2*synchronization_time in the past.

This is mainly needed for the TVB initial condition:
the [-min_tvb_delay, -min_tvb_delay/2] is sent also to the TVB->spikeNet transformer and at time 0 towards spikeNet,
whereas [-min_tvb_delay/2, 0.0] is sent at time 0 to the TVB->spikeNet transformer

Comment (dionperd) [reply tvb_contrib/tvb/contrib/cosimulation/cosimulator.py:315]: Thanks for the comments.

I made the changes, and also made the last additions to the CoSimulator (relative_output_time_steps) and corrected the specific test.

All tests PASS now.

**Comment by dionperd** [comment]
Thanks for the comments.

I made the changes, and also made the last additions to the CoSimulator (relative_output_time_steps) and corrected the specific test.

All tests PASS now.

File: tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py (hunk 2 above)
Comment (dionperd) [reply tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py]: Sorry for not noticing it! I had to also correct in the CoSimulator, and this is where I got confused and forgot it!
Done now!
Tests still PASS.

**Review by lionelkusch — APPROVED** [review]

File: tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py [hunk 5]
```diff
@@ -27,7 +27,7 @@
 
 import numpy
 
-from tvb.basic.neotraits.api import HasTraits, Attr, NArray
+from tvb.basic.neotraits.api import HasTraits, Attr, NArray, Int
```
Comment (lionelkusch) [inline tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py:30]:
```suggestion
from tvb.basic.neotraits.api import HasTraits, Attr, NArray
```

**Review by lionelkusch — APPROVED** [review]
